pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,343 @@
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
+func (e *DockercfgController) serviceAccountAdded(obj interface{}) {
+	serviceAccount := obj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(serviceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
+func (e *DockercfgController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
+	newServiceAccount := newObj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(newServiceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceAccountDeleted reacts to a ServiceAccount deletion by deleting all corresponding ServiceAccountToken Secrets
+func (e *DockercfgController) serviceAccountDeleted(obj interface{}) {
+	serviceAccount, ok := obj.(*api.ServiceAccount)
+	if !ok {
+		// Unknown type. If we missed a ServiceAccount deletion, the
+		// corresponding secrets will be cleaned up during the Secret re-list
+		return
+	}
+
+	secrets, err := e.listDockercfgSecrets(serviceAccount)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	for _, secret := range secrets {
+		if err := e.client.Secrets(secret.Namespace).Delete(secret.Name); err != nil {
+			glog.Errorf("Error deleting secret %s/%s: %v", secret.Namespace, secret.Name, err)
+		}
+	}
+}
+
+func getDockercfgSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-dockercfg-"
+}
+func getTokenSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-token-"
+}
+
+// createDockercfgSecretIfNeeded makes sure at least one ServiceAccountToken secret exists, and is included in the serviceAccount's Secrets list
+func (e *DockercfgController) createDockercfgSecretIfNeeded(serviceAccount *api.ServiceAccount) error {
+
+	// look for an ImagePullSecret in the form
+	dockercfgSecretName := ""
+	foundDockercfgImagePullSecret := false
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		if strings.HasPrefix(pullSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgImagePullSecret = true
+			dockercfgSecretName = pullSecret.Name
+			break
+		}
+	}
+	foundDockercfgMountableSecret := false
+	for _, mountableSecret := range serviceAccount.Secrets {
+		if strings.HasPrefix(mountableSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgMountableSecret = true
+			dockercfgSecretName = mountableSecret.Name
+			break
+		}
+	}
+
+	switch {
+	// if we already have a docker pull secret, simply return
+	case foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
+		return nil
+
+	case foundDockercfgImagePullSecret && !foundDockercfgMountableSecret, !foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
```

Comment:
if you insist, at least shorten the names to `foundPullSecret` and `foundMountableSecret`